client/src/components/Draftail/CommentableEditor/CommentableEditor.js

```diff
@@ -0,0 +1,217 @@
+
+function usePrevious(value) {
+  const ref = useRef();
+  useEffect(() => {
+    ref.current = value;
+  }, [value]); // Only re-run if value changes
+  return ref.current;
+}
+
+class DraftailInlineAnnotation {
+    constructor(field) {
+      this.field = field;
+      this.inlineRefs = new Set()
+    }
+    addRef(ref) {
+      this.inlineRefs.add(ref);
+    }
+    removeRef(ref) {
+      this.inlineRefs.delete(ref);
+    }
+    getDesiredPosition() {
+      const nodeTops = Array.from(this.inlineRefs).map(ref => ref.current).filter(node => node).map(node => node.getBoundingClientRect().top)
+      if (nodeTops.length > 0) {
+        return nodeTops.reduce((a,b) => a + b, 0)/nodeTops.length + document.documentElement.scrollTop
+      }
+      const fieldNode = this.field
+      if (fieldNode) {
+        return fieldNode.getBoundingClientRect().top + document.documentElement.scrollTop
+      }
+      return 0
+    }
+}
+
+function getFullSelectionState(contentState) {
+  // Get a selection state corresponding to the full contentState
+  const lastBlock = contentState.getLastBlock();
+  let fullSelectionState = SelectionState.createEmpty();
+  fullSelectionState = fullSelectionState.set('anchorKey', contentState.getFirstBlock().getKey());
+  fullSelectionState = fullSelectionState.set('focusKey', lastBlock.getKey());
+  fullSelectionState = fullSelectionState.set('anchorOffset', 0);
+  fullSelectionState = fullSelectionState.set('focusOffset', lastBlock.getLength());
+  return fullSelectionState
+}
+
+function getCommentControl(commentApp, contentPath, fieldNode) {
+  return ({getEditorState, onChange}) => {
+    return <ToolbarButton
+      name='comment'
+      active={false}
+      title={STRINGS.ADD_A_COMMENT}
+      icon={<Icon name="comment"/>}
+      onClick={() => {
+        const annotation = new DraftailInlineAnnotation(fieldNode)
+        const commentId = commentApp.makeComment(annotation, contentPath);
+        onChange(RichUtils.toggleInlineStyle(getEditorState(), COMMENT_STYLE_IDENTIFIER + commentId))
+      }}
+    />
+  }
+}
+
+function findCommentStyleRanges(contentBlock, callback) {
+    contentBlock.findStyleRanges((metadata) => metadata.getStyle().some((style) => style.startsWith(COMMENT_STYLE_IDENTIFIER)), (start, end) => {callback(start, end)})
+}
+
+function getCommentDecorator(commentApp) {
+  const CommentDecorator = ({ contentState, children }) => {
+    // The comment decorator makes a comment clickable, allowing it to be focused.
+    // It does not provide styling, as draft-js imposes a 1 decorator/string limit, which would prevent comment highlights
+    // going over links/other entities
+    const blockKey = children[0].props.block.getKey()
+    const start = children[0].props.start
+
+    const commentId = useMemo(() => parseInt(contentState.getBlockForKey(blockKey).getInlineStyleAt(start).find((style) => style.startsWith(COMMENT_STYLE_IDENTIFIER)).slice(8)), [blockKey, start]);
```

Comment:
Hmm this looks dangerous, but indeed I wouldn’t be too surprised if it was a necessary hack.

Reply:
I couldn't find another way of getting the offset (and thus style) in the decorator in the current draft version - it looks like the next version provides the start in the props, but unfortunately we're not there yet! Once we update in Draftail, this is a hack we can remove (but if you can find an alternative to this hack in the short term, I'd love to hear it of course, that would be great)